src/context/auth.tsx

```diff
@@ -0,0 +1,117 @@
+import React, { useContext, useReducer, useCallback } from "react";
+import axios from "axios";
+import { getUiUrl } from "utils/getEnvironmentVariables";
+
+interface State {
+  isAuthenticated: boolean;
+}
+
+const defaultState: State = {
+  isAuthenticated: false
+};
+
+type Action = { type: "authenticate" } | { type: "deauthenticate" };
+
+type Dispatch = (action: Action) => void;
+
+export type Logout = () => void;
+
+interface DispatchContext {
+  login: (LoginParams) => void;
+  logout: Logout;
+}
+
+const reducer = (state: State, action: Action): State => {
+  switch (action.type) {
+    case "authenticate":
+      return {
+        ...state,
+        isAuthenticated: true
+      };
+    case "deauthenticate":
+      return {
+        ...state,
+        isAuthenticated: false
+      };
+    default:
+      return state;
+  }
+};
+
+const logout = async (dispatch: Dispatch) => {
+  try {
+    await axios.get(`${getUiUrl()}/logout`);
+    dispatch({ type: "deauthenticate" });
+  } catch (error) {
+    // TODO: log errors
+  }
+};
+
+interface LoginParams {
+  username: string;
+  password: string;
+}
+const login = async (
+  dispatch: Dispatch,
+  { username, password }: LoginParams
+) => {
+  try {
+    await axios.post(`${getUiUrl()}/login`, { username, password });
+    dispatch({ type: "authenticate" });
+  } catch (error) {
+    // TODO: log errors
+  }
+};
+
+const AuthDispatchContext = React.createContext<DispatchContext | null>(null);
+const AuthStateContext = React.createContext<State | null>(null);
+
+const AuthProvider: React.FC = ({ children }) => {
+  const [state, dispatch] = useReducer(reducer, defaultState);
+
+  const logoutHandler = useCallback(() => {
+    logout(dispatch);
+  }, [dispatch]);
+
+  const loginHandler = useCallback(
+    ({ username, password }: LoginParams) => {
+      login(dispatch, { username, password });
+    },
+    [dispatch]
+  );
+
+  const dispatchContext: DispatchContext = {
+    login: loginHandler,
+    logout: logoutHandler
+  };
+
+  return (
+    <AuthDispatchContext.Provider value={dispatchContext}>
```

Comment:
What was your reasoning for having two separate contexts for AuthDispatch and AuthState? I think the information they provide is very similar and could be grouped into one context to make all auth related information simpler to access for consumers.

Reply:
read this article from kent c dodds: https://kentcdodds.com/blog/how-to-use-react-context-effectively

the value of these contexts is split into state and the dispatchers that update the state. Every time the value of context changes, all consumers are re-rendered. therefore, if a component only need to consume a dispatcher from auth context, then it should not re-render when the state is updated. splitting context into state and dispatch prevents these unnecessary re-renders.